oio/admin/agent.py

```diff
@@ -0,0 +1,570 @@
+        self.log.debug(msg % args)
+
+
+class OioAdminAgentClassManager(object):
+    """
+    Pluggable class manager for admin agent
+    """
+
+    def __init__(self, ctx=dict()):
+        self.blob_mover = BlobMoverWorker
+        self.blob_rebuilder = BlobRebuilder
+        self.conscience = ConscienceClient
+        self.meta2_mover = Meta2Database
+
+        for name, class_ in ctx.items():
+            setattr(self, name, class_)
+
+
+class OioAdminAgent(object):
+    """
+    Main class for the admin agent
+    Responsible of running admin tasks via multiprocessing
+    """
+    jobs = dict()
+
+    def __init__(self, args, **kwargs):
+        # This allows for easily swappable classes to be used
+        ctx = {}
+        try:
+            ctx = args.get("ctx", {})
+        except Exception:
+            pass
+        self.cm = OioAdminAgentClassManager(ctx=ctx or kwargs.get("ctx", {}))
+        self.host = args.host
+        self.namespace = args.namespace
+        self.client = self.cm.conscience({"namespace": args.namespace})
+        self.log = get_logger_from_args(args)
+        self.log.info("Starting oio-mover-agent")
+        self.jobs = dict()
+
+        signal.signal(signal.SIGTERM, self._clean_exit)
+        signal.signal(signal.SIGTERM, self._clean_exit)
+
+    def _clean_exit(self, signum, frame):
+        self.log.info(
+            "Clean exit requested: cleaning %d running jobs",
+            len([j for j in self.jobs.values() if j.get('status') == 0])
+        )
+        for job in self.jobs.values():
+            self._clean_stop(job)
+        self.log.info("Job cleanup completed. Shutting down")
+        sys.exit(0)
+
+    def _clean_stop(self, job):
+        if job.get('type') == 'meta2':
+            job['control'].get('signal').value = True
+            for p in job['processes']:
+                try:
+                    p.join()
+                except AssertionError:
+                    pass
+
+        elif job.get('type') == 'rawx':
+            for p in job['processes']:
+                try:
+                    p.terminate()
+                except Exception:
+                    pass
+
+    def _terminate(self, job):
+        for p in job['processes']:
+            try:
+                p.terminate()
+            except Exception:
+                pass
+
+    def move_meta2(self, config, stats, control):
+        """
+        Job for meta2 mover
+        In:
+        - config
+        - stats
+        - control
+        """
+        def _set(lock_, field, value):
+            lock_.acquire()
+            field.value = value
+            lock_.release()
+
+        def _add(lock_, field, value):
+            lock_.acquire()
+            field.value += value
+            lock_.release()
+
+        lock = control.get('lock')
+        src = config.get('src')
+        conf_str = ",".join(["%s=%s" % (k, v) for k, v in config.items()])
+        self.client.lock_score(dict(type="meta2", addr=src))
```

Comment:
Declaring dictionaries this way make things harder to read. I suggest this syntax: `{'type': 'meta2', 'addr': 'src'}`.